cpp/src/io/json/json_column.cu

```diff
@@ -0,0 +1,811 @@
+    rmm::exec_policy(stream),
+    unique_col_ids.begin(),
+    unique_col_ids.end(),
+    max_row_offsets.begin(),
+    [column_categories = column_categories.begin(),
+     parent_col_ids    = parent_col_ids.begin(),
+     max_row_offsets   = max_row_offsets.begin()] __device__(size_type col_id) {
+      auto parent_col_id = parent_col_ids[col_id];
+      while (parent_col_id != parent_node_sentinel and
+             column_categories[parent_col_id] != node_t::NC_LIST) {
+        col_id        = parent_col_id;
+        parent_col_id = parent_col_ids[parent_col_id];
+      }
+      return max_row_offsets[col_id];
+    },
+    [column_categories = column_categories.begin(),
+     parent_col_ids    = parent_col_ids.begin()] __device__(size_type col_id) {
+      auto parent_col_id = parent_col_ids[col_id];
+      return parent_col_id != parent_node_sentinel and
+             (column_categories[parent_col_id] != node_t::NC_LIST);
+      // Parent is not a list, or sentinel/root
+    });
+
+  return std::tuple{tree_meta_t{std::move(column_categories),
+                                std::move(parent_col_ids),
+                                std::move(column_levels),
+                                std::move(col_range_begin),
+                                std::move(col_range_end)},
+                    std::move(unique_col_ids),
+                    std::move(max_row_offsets)};
+}
+
+/**
+ * @brief Copies strings specified by pair of begin, end offsets to host vector of strings.
+ *
+ * @param input String device buffer
+ * @param node_range_begin Begin offset of the strings
+ * @param node_range_end End offset of the strings
+ * @param stream CUDA stream
+ * @return Vector of strings
+ */
+std::vector<std::string> copy_strings_to_host(device_span<SymbolT const> input,
+                                              device_span<SymbolOffsetT const> node_range_begin,
+                                              device_span<SymbolOffsetT const> node_range_end,
+                                              rmm::cuda_stream_view stream)
+{
+  auto const num_strings = node_range_begin.size();
+  rmm::device_uvector<thrust::pair<const char*, size_type>> string_views(num_strings, stream);
+  auto d_offset_pairs = thrust::make_zip_iterator(node_range_begin.begin(), node_range_end.begin());
+  thrust::transform(rmm::exec_policy(stream),
+                    d_offset_pairs,
+                    d_offset_pairs + num_strings,
+                    string_views.begin(),
+                    [input] __device__(auto const& offsets) {
+                      // Note: first character for non-field columns
+                      return thrust::make_pair(
+                        input.data() + thrust::get<0>(offsets),
+                        static_cast<size_type>(thrust::get<1>(offsets) - thrust::get<0>(offsets)));
+                    });
+  auto d_column_names = cudf::make_strings_column(string_views, stream);
+  auto to_host        = [](auto const& col) {
```

Comment:
nit: This (almost) IIFE seems redundant, if you return its result from the function again.

Reply:
This is mostly for readability, and modularity.